extract.c

```diff
@@ -0,0 +1,127 @@
+#include "squashfuse.h"
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include "squashfs_fs.h"
+
+#define PROGNAME "squashfuse_extract"
+
+#define ERR_MISC	(-1)
+#define ERR_USAGE	(-2)
+#define ERR_OPEN	(-3)
+
+static void usage() {
+    fprintf(stderr, "Usage: %s ARCHIVE PATH_TO_EXTRACT\n", PROGNAME);
+    fprintf(stderr, "       %s ARCHIVE -a\n", PROGNAME);
+    exit(ERR_USAGE);
+}
+
+static void die(const char *msg) {
+    fprintf(stderr, "%s\n", msg);
+    exit(ERR_MISC);
+}
+
+
+bool startsWith(const char *pre, const char *str)
+{
+    size_t lenpre = strlen(pre),
+    lenstr = strlen(str);
+    return lenstr < lenpre ? false : strncmp(pre, str, lenpre) == 0;
+}
+
+int main(int argc, char *argv[]) {
+    sqfs_err err = SQFS_OK;
+    sqfs_traverse trv;
+    sqfs fs;
+    char *image;
+    char *path_to_extract;
+    char *prefix;
+    char prefixed_path_to_extract[1024];
+    
+    prefix = "squashfs-root/";
+    
+    if(access(prefix, F_OK ) == -1 ) {
+        if (mkdir(prefix, 0777) == -1) {
+            perror("mkdir error");
+            exit(EXIT_FAILURE);
+        }
+    }
+    
+    if (argc != 3)
+        usage();
+    image = argv[1];
+    path_to_extract = argv[2];
+    
+    if ((err = sqfs_open_image(&fs, image, 0)))
+        exit(ERR_OPEN);
+    
+    if ((err = sqfs_traverse_open(&trv, &fs, sqfs_inode_root(&fs))))
+        die("sqfs_traverse_open error");
+    while (sqfs_traverse_next(&trv, &err)) {
+        if (!trv.dir_end) {
+            if((startsWith(path_to_extract, trv.path) != 0) || (strcmp("-a", path_to_extract) == 0)){
+                fprintf(stderr, "trv.path: %s\n", trv.path);
+                fprintf(stderr, "sqfs_inode_id: %lu\n", trv.entry.inode);
+                sqfs_inode inode;
+                if (sqfs_inode_get(&fs, &inode, trv.entry.inode))
+                    die("sqfs_inode_get error");
+                fprintf(stderr, "inode.base.inode_type: %i\n", inode.base.inode_type);
+                fprintf(stderr, "inode.xtra.reg.file_size: %lu\n", inode.xtra.reg.file_size);
+                strcpy(prefixed_path_to_extract, "");
+                strcat(strcat(prefixed_path_to_extract, prefix), trv.path);
+                if(inode.base.inode_type == SQUASHFS_DIR_TYPE){
+                    fprintf(stderr, "inode.xtra.dir.parent_inode: %ui\n", inode.xtra.dir.parent_inode);
```

Comment:
I could but then I would have to implement fancier arg parsing - do you know how to do that without bloating the code?